### PR TITLE
feature/products-subscribe

### PR DESCRIPTION
**Pull request by busticated**
## Description

Adds support for optional `--product <id>` flag to the `particle subscribe` command to listen for events on a product device ([ch](https://app.clubhouse.io/particle/story/48795/enable-subscribing-to-a-product-device-s-events))

## How to Test

_Note: requires you to have a Particle product with one or more devices_

Run `particle subscribe --help` and explore the command and options.

## Completeness

- [x] User is totes amazing for contributing!
- [x] Contributor has signed [CLA](https://docs.google.com/a/particle.io/forms/d/1_2P-vRKGUFg5bmpcKLHO_qNZWGi5HKYnfrrkd-sbZoA/viewform)
- [x] Problem and solution clearly stated
- [x] Tests have been provided
- [x] Docs have been updated
- [x] CI is passing

### CODE REVIEW

File: src/cmd/subscribe.js [hunk 1]
```diff
@@ -1,77 +1,90 @@
...
-					continue;
-				}
-				chunks.push(line);
-				if (line.indexOf('data:') === 0) {
-					processItem(chunks);
-					chunks = [];
-				}
+		if (product){
+			if (!device){
+				throw usageError(
```
Comment (monkbroc) [inline src/cmd/subscribe.js]: It's possible to subscribe all product events, not just product events from a specific device.

File: src/cmd/subscribe.js [hunk 2]
```diff
@@ -1,77 +1,90 @@
...
+	}
 
+	startListening({ device, all, product, params: { event } }){
 		// legacy argument order
 		let eventName = event[0];
 		let deviceId = event[1] || device;
 
 		// if they typed: "particle subscribe mine"
-		if (eventName === 'mine') {
+		if (eventName === 'mine'){
```
Comment (monkbroc) [inline src/cmd/subscribe.js]: It's time to remove the now undocumented use case of `particle subscribe mine`. That has been the default behavior for some time.

File: src/cmd/subscribe.js [hunk 3]
```diff
@@ -1,77 +1,90 @@
...
-				}
-			}
-
-			console.log(JSON.stringify(obj));
+		if (!deviceId){
+			this.ui.stdout.write(`Subscribing to ${eventLabel} from the firehose (all devices) and my personal stream (my devices)${os.EOL}`);
+		} else if (deviceId === 'mine'){
+			this.ui.stdout.write(`Subscribing to ${eventLabel} from my personal stream (my devices only)${os.EOL}`);
+		} else {
+			this.ui.stdout.write(`Subscribing to ${eventLabel} from ${deviceId}'s stream${os.EOL}`);
```
Comment (monkbroc) [inline src/cmd/subscribe.js]: There should be another line about subscribing to product events.

File: src/cmd/subscribe.js (hunk 1 above)
Comment (busticated) [reply src/cmd/subscribe.js]: ✅ done in https://github.com/particle-iot/particle-cli/pull/558/commits/ae3d983343e4e562175efc2587241e5e8f6062ec

File: src/cmd/subscribe.js (hunk 2 above)
Comment (busticated) [reply src/cmd/subscribe.js]: hehe! with pleasure 😉 

✅ done in https://github.com/particle-iot/particle-cli/pull/558/commits/ae3d983343e4e562175efc2587241e5e8f6062ec

File: src/cmd/subscribe.js (hunk 3 above)
Comment (busticated) [reply src/cmd/subscribe.js]: ✅ done in https://github.com/particle-iot/particle-cli/pull/558/commits/ae3d983343e4e562175efc2587241e5e8f6062ec

**Comment by busticated** [comment]
@monkbroc not sure how to test this one 👉 https://github.com/particle-iot/particle-cli/pull/558/commits/ae3d983343e4e562175efc2587241e5e8f6062ec#diff-b2e261f7d33882de918561ccc4318a84R63

**Comment by busticated** [comment]
@monkbroc ok, i think this should do it - here's a breakdown of the handled calls:

```sh
particle subscribe
# API ARGS: device: 'mine', event: undefined, product: undefined

particle subscribe <event-name>
# API ARGS: device: 'mine', event: <event-name>, product: undefined

particle subscribe --all <event-name>
# API ARGS: device: 'mine', event: <event-name>, product: undefined

particle subscribe --all
# USAGE ERROR: `event` parameter is required when `--all` flag is set

particle subscribe --device <device>
# API ARGS: device: <device>, event: undefined, product: undefined

particle subscribe --device <device> <event-name>
# API ARGS: device: <device>, event: <event-name>, product: undefined

particle subscribe --product <product-id>
# API ARGS: device: undefined, event: undefined, product: <product-id>

particle subscribe --product <product-id> <event-name>
# API ARGS: device: undefined, event: <event-name>, product: <product-id>

particle subscribe --product <product-id> --device <device>
# API ARGS: device: <device>, event: undefined, product: <product-id>

particle subscribe --product <product-id> --device <device> <event-name>
# API ARGS: device: <device>, event: <event-name>, product: <product-id>
```

**Pull request merged**